test/exe/admin_response_test.cc

```diff
@@ -0,0 +1,334 @@
+#include "test/exe/main_common_test_base.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+class AdminStreamingTest : public AdminRequestTestBase, public testing::Test {
+protected:
+  static constexpr absl::string_view StreamingEndpoint = "/stream";
+
+  class StreamingAdminRequest : public Envoy::Server::Admin::Request {
+  public:
+    static constexpr uint64_t NumChunks = 10;
+    static constexpr uint64_t BytesPerChunk = 10000;
+
+    StreamingAdminRequest(std::function<void()>& get_headers_hook,
+                          std::function<void()>& next_chunk_hook)
+        : chunk_(BytesPerChunk, 'a'), get_headers_hook_(get_headers_hook),
+          next_chunk_hook_(next_chunk_hook) {}
+    Http::Code start(Http::ResponseHeaderMap&) override {
+      get_headers_hook_();
+      return Http::Code::OK;
+    }
+    bool nextChunk(Buffer::Instance& response) override {
+      next_chunk_hook_();
+      response.add(chunk_);
+      return --chunks_remaining_ > 0;
+    }
+
+  private:
+    const std::string chunk_;
+    uint64_t chunks_remaining_{NumChunks};
+    std::function<void()>& get_headers_hook_;
+    std::function<void()>& next_chunk_hook_;
+  };
+
+  AdminStreamingTest() : AdminRequestTestBase(Network::Address::IpVersion::v4) {
+    startEnvoy();
+    started_.WaitForNotification();
+    Server::Admin& admin = *main_common_->server()->admin();
+    admin.addStreamingHandler(
+        std::string(StreamingEndpoint), "streaming api",
+        [this](Server::AdminStream&) -> Server::Admin::RequestPtr {
+          return std::make_unique<StreamingAdminRequest>(get_headers_hook_, next_chunk_hook_);
+        },
+        true, false);
+  }
+
+  using ChunksBytes = std::pair<uint64_t, uint64_t>;
+  ChunksBytes runStreamingRequest(AdminResponseSharedPtr response,
+                                  std::function<void()> chunk_hook = nullptr) {
+    absl::Notification done;
+    std::vector<std::string> out;
+    absl::Notification headers_notify;
+    response->getHeaders(
+        [&headers_notify](Http::Code, Http::ResponseHeaderMap&) { headers_notify.Notify(); });
+    headers_notify.WaitForNotification();
```

Comment:
I don't feel that strongly about it, but for now I'll leave a note indicating maybe a promise would be better.

Would a promise still be better if I add a test for a timeout and change the above 'wait' call to include one?

Reply:
Thanks for the example. I can see how that is nicer in some ways, but in this case since I am aggregating ResponseData (formerly ChunkBytes) across both the headers and nextChunk callbacks, I think it wouldn't change much. It would actually be a direct replacement, line-for-line.

I don't object to use of promise for this sort of thing but given the existing use of notifications elsewhere in the pre-existing AdminRequest infrastructure in a few places, and the timeout, I'd prefer to stick with one mechanism.